sensorprocess/movementsensorprocess.go

```diff
@@ -13,85 +13,131 @@ import (
+// cartofacade.
+func (config *Config) addMovementSensorReadingInOnline(ctx context.Context) error {
+	// get next movement sensor data response; ignoring status since it is always false
+	movementSensorReading, err := config.MovementSensor.TimedMovementSensorReading(ctx)
 	if err != nil {
-		config.Logger.Warn(err)
 		if errors.Is(err, replaymovementsensor.ErrEndOfDataset) {
 			time.Sleep(1 * time.Second)
 		}
 		return err
 	}
 
-	// add IMU data to cartographer and sleep remainder of time interval
-	timeToSleep := config.tryAddIMUReadingOnce(ctx, *imuReading.TimedIMUResponse)
+	// add movement sensor data to cartographer and sleep remainder of time interval
+	timeToSleep := config.tryAddMovementSensorReadingOnce(ctx, movementSensorReading)
 
-	if !imuReading.TestIsReplaySensor {
+	if !movementSensorReading.TestIsReplaySensor {
 		time.Sleep(time.Duration(timeToSleep) * time.Millisecond)
-		config.Logger.Debugf("imu sleep for %vms", timeToSleep)
+		config.Logger.Debugf("movement sensor sleep for %vms", timeToSleep)
 	}
 
 	return nil
 }
 
-// tryAddIMUReadingUntilSuccess adds a reading to the cartofacade and retries on error (offline mode).
+// tryAddMovementSensorReadingUntilSuccess adds a reading to the cartofacade and retries on error (offline mode).
 // While add sensor reading fails, keep trying to add the same reading - in offline mode we want to
 // process each reading so if we cannot acquire the lock we should try again.
-func (config *Config) tryAddIMUReadingUntilSuccess(ctx context.Context, reading s.TimedIMUReadingResponse) error {
+func (config *Config) tryAddMovementSensorReadingUntilSuccess(ctx context.Context, reading s.TimedMovementSensorReadingResponse) error {
+	var imuDone, odometerDone bool
+	if !config.MovementSensor.Properties().IMUSupported {
+		imuDone = true
+	}
+	if !config.MovementSensor.Properties().OdometerSupported {
+		odometerDone = true
+	}
 	for {
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
 		default:
-			if err := config.tryAddIMUReading(ctx, reading); err != nil {
-				if !errors.Is(err, cartofacade.ErrUnableToAcquireLock) {
-					config.Logger.Warnw("Retrying sensor reading due to error from cartofacade", "error", err)
+			if !odometerDone {
+				if err := config.tryAddOdometerReading(ctx, *reading.TimedOdometerResponse); err != nil {
+					if !errors.Is(err, cartofacade.ErrUnableToAcquireLock) {
+						config.Logger.Warnw("Retrying odometer sensor reading due to error from cartofacade", "error", err)
+					}
+				} else {
+					odometerDone = true
 				}
-			} else {
+			}
+			if !imuDone {
+				if err := config.tryAddIMUReading(ctx, *reading.TimedIMUResponse); err != nil {
+					if !errors.Is(err, cartofacade.ErrUnableToAcquireLock) {
+						config.Logger.Warnw("Retrying IMU sensor reading due to error from cartofacade", "error", err)
+					}
+				} else {
+					imuDone = true
+				}
+			}
+			if imuDone && odometerDone {
```

Comment:
[q] Do we want to stop only if both are done, or if either is done?

Reply:
I think this is covered by lines 58-64, where she's setting them to done if the properties are not supported, but maybe a comment to explain this kat.

Reply:
What about the instance of an offline run where, during data generation, the odom got disconnected but the movement sensor kept gathering data for some time? Waiting for both may cause cartographer to attempt to add 1000s of IMU data points after odom data is complete.

Another example of this could be if a user is lax in their time interval definition and accidentally grabs data from a different run which only had an odom movement sensor

Reply:
@randhid for context, we had this issue in the past where leftover imu data would cause the session to hang until timeout occurred.

After running through the full pipeline for offline mode, I believe this is handled by the new offline architecture, but wanted to make you aware of it in case it crops up again

Reply:
I don't think this is going to be a problem. Since we support using both odom and imu together only via merged movement sensors, if one runs out of data, we get an`ErrEndOfDataset` error.  (see [`TimedMovementSensorReading`](https://github.com/kkufieta/viam-cartographer/blob/kk/rsdk-4862-sensorprocess-2/sensors/movementsensor.go#L95). If that happens, we [exit out](https://github.com/kkufieta/viam-cartographer/blob/kk/rsdk-4862-sensorprocess-2/sensorprocess/sensorprocess.go#L157) and don't call this function.